### PR TITLE
Added Dataset.py WIP

### PR DESCRIPTION
**Pull request by Z-Nouman**

### CODE REVIEW

File: dumbbells/dataset.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        # For the environment "MountainCar-v0", there are 2 observations, presented in the following order:
+        # Car Position: Ranges from -1.2 to 0.6
+        # Car Velocity: Ranges from -0.07 to 0.07
+        # The car starts at some position between -0.6 to -0.4
+        self.state_space = self.env.reset()
+
+        # For the environment "MountainCar-v0", there are 2 possible rewards:
+        # Reward of 0 is awarded if the agent has reached the flag (position 0.5)
+        # Reward of -1 is awarded if the agent is at a position < 0.5
+        self.reward_space = [-1, 0]
```
Comment (DonatasTamosauskas) [inline dumbbells/dataset.py:41]: Is the reward space the same for all environments, or could we somehow get it out of the `env` itself?

File: dumbbells/dataset.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        prev_state = self.get_state()
+        # Take the action
+        result = self.env.step(action)
+        # Update the state and memory
+        self.state_space = result[0]
+        self.pushMem(prev_state, action, result[1], self.get_state())
+        # Return the result of the action
+        return result
+
+    def pushMem(self, prev_state, action, reward, next_state):
```
Comment (DonatasTamosauskas) [inline dumbbells/dataset.py]: The pythonic variable naming standard uses all lowercase and `_`, so the name would be `push_mem` or `pushmem`, unless this is a convention form some library that I'm not aware of

File: dumbbells/dataset.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            reward: Reward given for taking that action at that state
+            next_state: Current state as a result of taking that action
+
+        Return: None
+        """
+        if len(self.memory) < self.memory_size:
+            self.memory.append(None)
+        self.memory[self.position] = (prev_state, action, reward, next_state)
+        self.position = (self.position + 1) % self.memory_size
+        return
```
Comment (DonatasTamosauskas) [inline dumbbells/dataset.py]: `return` statement is not needed if nothing is returned

File: tests/test_dataset.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+def test_memory():
+    """Tests the __len__() and __getitem__() methods"""
+
+    game = "MountainCar-v0"
+    memory_size = 5
+    dataset = Dataset(game, memory_size)
+    dataset.step(2)
+    prev_state = dataset.get_state()
+
+    assert dataset.__len__() == 1
```
Comment (DonatasTamosauskas) [inline tests/test_dataset.py]: Here you can simply call the `len` method, so: `assert len(dataset) == 1`

File: tests/test_dataset.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    result = dataset.step(0)
+    image = dataset.__getitem__(1)
```
Comment (DonatasTamosauskas) [inline tests/test_dataset.py]: Similar story here, it's enough to index it: `image = dataset[1]`

File: tests/test_dataset.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    assert image[0][0] == prev_state[0]
+    assert image[0][1] == prev_state[1]
+    assert image[1] == 0
+    assert image[2] == result[1]
+    assert image[3][0] == result[0][0]
+    assert image[3][1] == result[0][1]
+    assert dataset.__len__() == 2
+
+
+if __name__ == "__main__":
```
Comment (DonatasTamosauskas) [inline tests/test_dataset.py]: Since we're using `pytest` for tests, there's no need to run the tests manually, so no need to include this part.

**Review by DonatasTamosauskas — CHANGES_REQUESTED** [review]
Very clean code, I like it!!

File: dumbbells/dataset.py (hunk 1 above)
Comment (Adriman32) [reply dumbbells/dataset.py:41]: I don't think that gym allows for extracting the reward. From what I saw in the documentation and the GitHub for mountain_car.py, it was just a float that is returned, either -1 or 0. There isn't really a variable we can pull from that I could see.

**Review by DonatasTamosauskas — APPROVED** [review]

**Pull request merged**